File: en/codes/typescript/chapter_searching/two_sum.ts
/**
 * File: two_sum.ts
 * Created Time: 2022-12-15
 * Author: gyt95 (gytkwan@gmail.com)
 */

/* Method one: Brute force enumeration */
function twoSumBruteForce(nums: number[], target: number): number[] {
    const n = nums.length;
    // Two-layer loop, time complexity is O(n^2)
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }
    return [];
}

/* Method two: Auxiliary hash table */
function twoSumHashTable(nums: number[], target: number): number[] {
    // Auxiliary hash table, space complexity is O(n)
    let m: Map<number, number> = new Map();
    // Single-layer loop, time complexity is O(n)
    for (let i = 0; i < nums.length; i++) {
        let index = m.get(target - nums[i]);
        if (index !== undefined) {
            return [index, i];
        } else {
            m.set(nums[i], i);
        }
    }
    return [];
}

/* Driver Code */
// Method one
const nums = [2, 7, 11, 15],
    target = 13;

let res = twoSumBruteForce(nums, target);
console.log('Method one res =', res);

// Method two
res = twoSumHashTable(nums, target);
console.log('Method two res =', res);

export {};

File: en/codes/rust/chapter_searching/two_sum.rs
/*
 * File: two_sum.rs
 * Created Time: 2023-01-14
 * Author: xBLACICEx (xBLACKICEx@outlook.com), codingonion (coderonion@gmail.com)
 */

include!("../include/include.rs");

use std::collections::HashMap;

/* Method one: Brute force enumeration */
pub fn two_sum_brute_force(nums: &Vec<i32>, target: i32) -> Option<Vec<i32>> {
    let size = nums.len();
    // Two-layer loop, time complexity is O(n^2)
    for i in 0..size - 1 {
        for j in i + 1..size {
            if nums[i] + nums[j] == target {
                return Some(vec![i as i32, j as i32]);
            }
        }
    }
    None
}

/* Method two: Auxiliary hash table */
pub fn two_sum_hash_table(nums: &Vec<i32>, target: i32) -> Option<Vec<i32>> {
    // Auxiliary hash table, space complexity is O(n)
    let mut dic = HashMap::new();
    // Single-layer loop, time complexity is O(n)
    for (i, num) in nums.iter().enumerate() {
        match dic.get(&(target - num)) {
            Some(v) => return Some(vec![*v as i32, i as i32]),
            None => dic.insert(num, i as i32),
        };
    }
    None
}

fn main() {
    // ======= Test Case =======
    let nums = vec![2, 7, 11, 15];
    let target = 13;

    // ====== Driver Code ======
    // Method one
    let res = two_sum_brute_force(&nums, target).unwrap();
    print!("Method one res =");
    print_util::print_array(&res);
    // Method two
    let res = two_sum_hash_table(&nums, target).unwrap();
    print!("\nMethod two result res = ");
    print_util::print_array(&res);
}
